File: 1. Bubble Sort - GFG/1.-bubble-sort.cpp
// { Driver Code Starts
//Initial Template for C++

// C program for implementation of Bubble sort
#include <stdio.h>

// swapping the elements
void swap(int *xp, int *yp)
{
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}


 // } Driver Code Ends
//User function Template for C++

class Solution{
    public:
    void swap(int &a, int &b){ // helper function for swapping
        int temp = a;
               a = b;
               b = temp;
    }
    //Function to sort the array using bubble sort algorithm.
    void bubbleSort(int arr[], int n){
        // Optimised Bubble sort for best case of O(n) when there are no swaps at all in the pass
        int swapped = 1;
        for (int pass = n-1; pass >=0 && swapped; pass--){
            swapped = 0;
            for (int j = 0; j <= pass-1; j++){
                if(arr[j] > arr[j+1]){
                    swap(arr[j], arr[j+1]);
                    swapped = 1;
                }
            
            }
            if (swapped == 0){
                break;
            }
            }
            
        }
    
};


// { Driver Code Starts.

/* Function to print an array */
void printArray(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}
 
// Driver program to test above functions
int main()
{
    int arr[1000],n,T,i;
  
    scanf("%d",&T);
    
    while(T--){
        
    scanf("%d",&n);
    
    for(i=0;i<n;i++)
      scanf("%d",&arr[i]);
      
    Solution ob;  
      
    ob.bubbleSort(arr, n);
    printArray(arr, n);
    }
    return 0;;
}  // } Driver Code Ends

File: Sorting_DSA_Basics/1. Bubble Sort - GFG/1.-bubble-sort.py
#User function Template for python3

class Solution:
    #Function to sort the array using bubble sort algorithm.
    def bubbleSort(self,arr, n):
        # code here
        for i in range(n):                             #loop goes from 1st to last element 
            for j in range(n - i - 1):                 #loop goes form 1st to last unordered element
                if arr[j] > arr[j+1]:                  #that has bubbled till there by swappinng
                    arr[j],arr[j+1] = arr[j+1],arr[j]  # python swap syntax
    

        
    

#{ 
#  Driver Code Starts
#Initial Template for Python 3



if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ob.bubbleSort(arr, n)
        for i in arr:
            print(i,end=' ')
        print()

# } Driver Code Ends